Feature/shared ptr

panchock opened the pull request.

panchock committed:
shared_ptr

panchock committed:
shared_ptr

panchock committed:
docs

panchock committed:
delete noexcept
default startValue in Counter ctor

panchock committed:
line break fixes

panchock committed:
warning level 4

panchock committed:
remove line breaks

Thankjnv commented on Pointers/Counter.h:
```diff
@@ -0,0 +1,13 @@
+#pragma once
+#include <cstdint>
+class Counter {
```
Add a blank line between the includes and the class definition

Thankjnv commented on Pointers/SharedPtr.inl:
```diff
@@ -0,0 +1,83 @@
+			destroy();
+		}
+	} catch (...) {
+	}
```
Add a comment of `Intentionally Left Blank` in the empty block

Thankjnv commented on Pointers/SharedPtr.inl:
```diff
@@ -0,0 +1,83 @@
+}
+
+template<typename T>
+SharedPtr<T>& SharedPtr<T>::operator=(const SharedPtr<T>& other) {
```
What would happen in the following code block?
```cpp
SharedPtr<int> a(new int(5));
a = a;
```

Thankjnv commented on Pointers/Counter.cpp:
```diff
@@ -0,0 +1,34 @@
+	printState(); 
+}
+
+Counter::~Counter() {
```
Why do you need to implement a destructor here? Once you destroy the object no one can access m_counter

Thankjnv commented on Pointers/SharedPtr.inl:
```diff
@@ -0,0 +1,83 @@
+
+template<typename T>
+void SharedPtr<T>::destroy() {
+	if (m_ptr != nullptr && m_refCount->getCounter() == 0) {
```
If the original pointer passed to the SharedPtr is nullptr (which is a valid value) you are going to leak the refCount

Thankjnv commented on Pointers/SharedPtr.inl:
```diff
@@ -0,0 +1,83 @@
+}
+
+template<typename T>
+void SharedPtr<T>::destroy() {
```
It looks like whenever you call destroy you decrease the refcount directly before doing so
It would make more sense to decrease at the beginning of destroy instead of duplicate calls everywhere

Thankjnv reviewed (commented):
In addition, remove all of the prints before the next CR

panchock committed:
add line break

panchock committed:
remove destructor of Counter

panchock committed:
added make shared

panchock committed:
decrease refcount in destroy func
destructor catch ILB
fix self copy assignment
fix refcount leak in destroy func

panchock committed:
Merge branch 'main' into feature/shared_ptr

# Conflicts:
#	Pointers/UniquePtr.h

added make_unique

ExhoAR22 commented on Pointers/Counter.cpp:
```diff
@@ -0,0 +1,30 @@
+using std::endl;
+
+Counter::Counter(uint32_t startValue) : m_counter(startValue) {
+	cout << "Initialize refcount" << endl;
```
It's usually a bad idea to do I/O in the ctor, since you don't want to force output on people who might use your code in the future. I assume you added this in for debugging purposes - but it shouldn't be in the pull request. Please remove it.

panchock committed:
reset pointer to nullptr

panchock committed:
no debug prints

ExhoAR22 commented on Pointers/Counter.h:
```diff
@@ -0,0 +1,13 @@
+public:
+	Counter(uint32_t startValue = 1);
+	void decrease();
+	void increase();
```
How about overriding `operator++` and `operator--`?

ExhoAR22 commented on Pointers/SharedPtr.h:
```diff
@@ -0,0 +1,32 @@
+#include "Counter.h"
+
+using std::cout;
+using std::endl;
```
Don't have `using` statements in headers (especially not in the global namespace), it pollutes the namespace for others

panchock committed:
explicit ctor
get only rvalue in ctor

panchock committed:
remove more debug prints

ExhoAR22 commented on Pointers/SharedPtr.inl:
```diff
@@ -0,0 +1,90 @@
+
+template<typename T>
+SharedPtr<T>::SharedPtr(T* ptr) : m_ptr(ptr), m_refCount(new Counter) {
+	cout << "SharedPtr ctor" << endl;
```
Same here. I/O in the ctor/dtor is not a good practice. Please remove it.

panchock committed:
remove rvalue limitation from ctor

ExhoAR22 commented on Pointers/main.cpp:
```diff
@@ -1,25 +1,56 @@
+template<typename T, class... Args>
+UniquePtr<T> make_unique(Args&&... args) {
+	return UniquePtr<T>(new T(std::forward<Args>(args)...));
+}
```
Why are these functions here and not in their own header file?

ExhoAR22 reviewed (changes requested):
Nice! Please fix the few small issues Iv'e pointed out and you're done :)

panchock committed:
opeartor++ and operator--

panchock committed:
delete using from .h

panchock committed:
Utils.h for make_shared and make_unique
